<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Microbial Analysis - Microbial Sequencing Techniques</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../hexmedfly.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.0/css/all.min.css">


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../hexmedfly.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Microbial Analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../qmds/introduction.html" rel="" target="">
 <span class="menu-text">Overview</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-diversity-metrics-and-ordination-methods" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Diversity Metrics and Ordination Methods</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-diversity-metrics-and-ordination-methods">    
        <li>
    <a class="dropdown-item" href="../qmds/alpha.html" rel="" target="">
 <span class="dropdown-text">Alpha Diversity Analysis Methods</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../qmds/beta.html" rel="" target="">
 <span class="dropdown-text">Beta Diversity Analysis Methods</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../qmds/PCA.html" rel="" target="">
 <span class="dropdown-text">Principal Component Analysis (PCA)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../qmds/PCoA.html" rel="" target="">
 <span class="dropdown-text">Principal Co-ordinate Analysis (PCoA)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../qmds/NMDS.html" rel="" target="">
 <span class="dropdown-text">Non-Metric Multidimensional Scaling (NMDS)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../qmds/microbialsummary.html" rel="" target="">
 <span class="dropdown-text">The Microbial Summary</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-tephritid-microbiota" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Tephritid Microbiota</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-tephritid-microbiota">    
        <li>
    <a class="dropdown-item" href="../qmds/tephritidecology.html" rel="" target="">
 <span class="dropdown-text">Distribution of Tephritid Pests</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-profiling-the-microbiota" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Profiling the Microbiota</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-profiling-the-microbiota">    
        <li>
    <a class="dropdown-item" href="../qmds/sequencingtechniques.html" rel="" target="">
 <span class="dropdown-text">Microbial Sequencing Techniques</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../qmds/dnaextraction.html" rel="" target="">
 <span class="dropdown-text">DNA Extraction</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <div class="quarto-navbar-tools ms-auto">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#an-introduction-to-amplicon-sequencing" id="toc-an-introduction-to-amplicon-sequencing" class="nav-link active" data-scroll-target="#an-introduction-to-amplicon-sequencing">An Introduction to Amplicon Sequencing</a></li>
  <li><a href="#the-types-of-amplicon-sequencing" id="toc-the-types-of-amplicon-sequencing" class="nav-link" data-scroll-target="#the-types-of-amplicon-sequencing">The Types of Amplicon Sequencing</a>
  <ul class="collapse">
  <li><a href="#short-read-sequencing" id="toc-short-read-sequencing" class="nav-link" data-scroll-target="#short-read-sequencing">Short-read sequencing:</a></li>
  <li><a href="#long-read-sequencing" id="toc-long-read-sequencing" class="nav-link" data-scroll-target="#long-read-sequencing">Long-read sequencing:</a></li>
  <li><a href="#how-amplicon-sequencing-works" id="toc-how-amplicon-sequencing-works" class="nav-link" data-scroll-target="#how-amplicon-sequencing-works">How amplicon sequencing works:</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Microbial Sequencing Techniques</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<div class="cell" data-layout-align="left">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="images/sequencing.png" class="img-fluid figure-img" width="762"></p>
<figcaption class="figure-caption">This image was created on Canva.</figcaption>
</figure>
</div>
</div>
</div>
<p>Before we actually have the data to analyse, we must analyse the DNA of our samlples through microbial sequencing technqiues. This section will take you through the sequencing techniques used to anlayse microbial data.</p>
<section id="an-introduction-to-amplicon-sequencing" class="level2">
<h2 class="anchored" data-anchor-id="an-introduction-to-amplicon-sequencing">An Introduction to Amplicon Sequencing</h2>
<p>Amplicon sequencing is a targeted DNA sequencing approach used to identify and compare specific genetic regions across many samples. For bacteria and archaea, this typically focuses on the <strong>16S rRNA gene</strong>. &nbsp;<br>
&nbsp;</p>
<p><strong>What is the 16S rRNA gene?</strong></p>
<div class="cell" data-layout-align="left">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="images/16SrRNA.png" class="img-fluid figure-img" width="154"></p>
<figcaption class="figure-caption">Molecular structure of the 30S Subunit, Schluenzen et al.&nbsp;2000.</figcaption>
</figure>
</div>
</div>
</div>
<p>The 16S rRNA (ribosomal RNA) gene encodes the small subunit ribosomal RNA (SSU rRNA), which is a core structural and functional component of the 30S subunit of the prokaryotic ribosome. This RNA plays an essential role in protein synthesis.<br>
&nbsp;<br>
&nbsp;</p>
<p><strong>What does ‘16S’ mean?</strong></p>
<p>The term <em>16S</em> refers to the sedimentation coefficient of the RNA molecule, the sedimentation coefficient refers to the rate at which a particle sediments in a centrifugal field. It is measured in Svedberg units, rather than its physical size or molecular weight. This means there are 16 Svedberg units. &nbsp;<br>
&nbsp;</p>
<p><strong>Why is the 16S rRNA gene used in microbiota studies?</strong></p>
<p>The 16S rRNA gene is widely used in microbiota studies because it is highly conserved across bacterial and archaeal taxa, meaning its sequence has remained very similar across different species over evolutionary time. This conservation allows for broad amplification of the gene using universal primers, while the presence of variable regions enables taxonomic discrimination between microorganisms.</p>
<p>&nbsp;<br>
&nbsp;<br>
<strong>What primers are used to sequence the 16S rRNA gene?</strong></p>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/primers.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption class="figure-caption">The forward and reverse primers of the 16S rRNA gene. THis image was generated on Canva.</figcaption>
</figure>
</div>
</div>
</div>
<p>The primer pair used which targets the full-length 16S rRNA gene is 27F and 1492R, sequencing approaches that capture this entire gene are typically referred to as <strong>long-read sequencing</strong>. &nbsp;<br>
&nbsp;</p>
<p><strong>What are hypervariable regions, and conserved regions, and why are they important?</strong></p>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/regions.png" class="img-fluid figure-img" style="width:80.0%"></p>
<figcaption class="figure-caption">The hypervariable (V1-V9), and conserved regions of the 16SrRNA gene. This image was created on Canva.</figcaption>
</figure>
</div>
</div>
</div>
<p>The <strong>16S rRNA gene (~1,500 bp)</strong> is evolutionary constrained, but it is interspersed with regions that are able to tolerate more sequence variation.</p>
<p>This creates a mosaic like stucture of:</p>
<ul>
<li><p><strong>9 hypervariable regions (V1-V9)</strong></p></li>
<li><p><strong>Intervening conserved (constant) regions</strong></p></li>
</ul>
<p>The <strong>hypervariable regions</strong> can provide species- or genus-specific signature sequences for bacterial identification. These hypervariable regions range from approximately 30 to 100 base pairs in length, and their degree of conservation varies widely. Sequencing the full-length 16S rRNA gene (approximately 1,500 bp) enables comparison across all 9 hypervariable regions, improving taxonomic resolution.</p>
<p><strong>Conserved regions</strong> on the other hand, are regions that have remained relatively unchanged across bacteria over long evolutionary periods due to strong selective pressure. As a result, these regions evolve slowly and retain high sequence similarity even among distantly related bacteria. They also have an important function of being criticial for ribosome structure and protein synthesis, which is essential for bacterial survival.<br>
</p>
<p>In short, <strong>conserved regions</strong> are the universal primers, <strong>hypervariable regions</strong> allow for taxonomic discrimination.<br>
</p>
<p>&nbsp;</p>
</section>
<section id="the-types-of-amplicon-sequencing" class="level2">
<h2 class="anchored" data-anchor-id="the-types-of-amplicon-sequencing">The Types of Amplicon Sequencing</h2>
<p>Within bacterial amplicon sequencing, there are two main types:</p>
<section id="short-read-sequencing" class="level3">
<h3 class="anchored" data-anchor-id="short-read-sequencing">Short-read sequencing:</h3>
<p><br>
Produces high-accuracy reads of relatively short DNA fragments, commonly used for microbial community profiling.</p>
<p>It is referred to as <strong>short-read sequencing</strong> because it only sequences a section of the <strong>16S rRNA gene</strong>. In short read sequences only 1-2 specific hypervariable regions will be selected. There are a range of hypervariable regions commonly used:</p>
<ul>
<li><u><strong>V1 - V3</strong>:</u> <em>27-534 bp</em> of the 16S gene.<br>
</li>
<li><u><strong>V3</strong>:</u> <em>341-534 bp</em> of the 16S gene.<br>
</li>
<li><u><strong>V3 - V4</strong>:</u> <em>341-806 bp</em> of the 16S gene.<br>
</li>
<li><u><strong>V4</strong>:</u> <em>515-806 bp</em> of the 16S gene.</li>
</ul>
<p><strong>Short-read sequencing</strong> is typically performed using high-throughput sequencing platforms, these are designed to generate millions of reads in parallel. The <strong>Illumina</strong> sequencing platform has become the dominant technology for <strong>short read 16S rRNA gene amplicon sequencing</strong>.</p>
<p><strong>Illumina Technology</strong></p>
<ul>
<li><p>Illumina uses sequencing-by-synthesis (SBS) chemistry. This includes fluorescently labelled, reversible terminator nucleotides being incorporated one base at a time and imaged after each cycle.</p></li>
<li><p>Illuimina is known for havig high accuracy: Illumina platforms typically achieve ≥99.9% per-base accuracy, making them a gold standard for applications where low error rates are critical (e.g.&nbsp;variant calling, microbial community profiling).</p></li>
<li><p>Depending on the instrument used (e.g.&nbsp;MiSeq, NextSeq, NovaSeq), Illumina can generate millions to billions of reads per run, enabling large-scale population studies and deep metagenomic sequencing.</p></li>
<li><p>The cost per base is very low.</p></li>
</ul>
</section>
<section id="long-read-sequencing" class="level3">
<h3 class="anchored" data-anchor-id="long-read-sequencing">Long-read sequencing:</h3>
<p><br>
Produces longer DNA reads, allowing better resolution of complex genomic regions and full-length 16S rRNA gene sequencing.</p>
<p>Sequencing platforms for high-throughput sequencing: <strong>Pacific Biosciences (PacBio)</strong> Uses single-molecule real-time (SMRT) sequencing to generate highly accurate long reads (HiFi reads), well suited for full-length 16S rRNA sequencing and genome assembly.</p>
<p><strong>Oxford Nanopore Technologies (ONT)</strong> Sequences DNA by measuring changes in electrical current as molecules pass through nanopores; enables ultra-long reads and real-time sequencing, though with higher raw error rates than short-read platforms.</p>
</section>
<section id="how-amplicon-sequencing-works" class="level3">
<h3 class="anchored" data-anchor-id="how-amplicon-sequencing-works">How amplicon sequencing works:</h3>
<ol type="1">
<li>DNA extraction from samples is conducted (e.g.&nbsp;gut, soil, feces, fruit, insect tissue)</li>
</ol>
<p>Samples such as gut contents, soil, feces, fruit, insect tissue, etc, are first collected. DNA is extracted from these samples using a commercial DNA extraction kit or laboratory protocol. This process breaks open cells and isolates total DNA, which is then purified and stored in a solution (the DNA extract) for downstream analysis.</p>
<ol start="2" type="1">
<li>In the lab, PCR amplification is used to target a specific gene region using primers that bind conserved sites.</li>
</ol>
<p>Once a DNA sample has been extracted, a PCR reaction mix is prepared. This typically includes:</p>
<ul>
<li>The extracted DNA as a template<br>
</li>
<li>Forward and reverse primers (chosen based on short-read or long-read sequencing)<br>
</li>
<li>DNA polymerase, nucleotides (dNTPs), buffer, and other necessary solutions</li>
</ul>
<p>The PCR reaction is then carried out using standard thermal cycling steps:</p>
<ul>
<li><strong>Initial denaturation</strong> – separates the DNA strands<br>
</li>
<li><strong>Denaturation</strong> – repeated in cycles to keep strands separate<br>
</li>
<li><strong>Annealing</strong> – primers bind to the target regions<br>
</li>
<li><strong>Extension</strong> – DNA polymerase synthesizes new DNA</li>
</ul>
<p>These steps are repeated for multiple cycles to amplify the target region.</p>
<ol start="3" type="1">
<li>Indexing/barcoding so many samples can be pooled together.</li>
</ol>
<p>Indexing (barcoding) is the process of adding a short, unique DNA sequence (an index or barcode) to each sample during library preparation. Because each sample gets a different index, many samples can be pooled together and sequenced in a single run. After sequencing, the reads are demultiplexed—sorted back to their original samples—based on these index sequences.</p>
<ol start="4" type="1">
<li>High-throughput sequencing.</li>
</ol>
<p>Depending on whether short-read or long-read sequencing is used, high-throughput sequencing is used.</p>
<ol start="5" type="1">
<li>Bioinformatic analysis to infer taxonomic identity or genetic variation.</li>
</ol>
<p>A pipeline is used to analyse the data.</p>
<hr>
<p><strong>What does 16S actually mean in the 16S rRNA gene?</strong></p>
<div>
<form id="quiz-form-2">
<p><input type="radio" name="q2" value="a"> A. The gene is exactly 16,000 base pairs long<br>
<br></p>
<p><input type="radio" name="q2" value="b"> B. 16S refers to the sedimentation coefficient, meaning the rate at which the particle sediments in a centrifugal field<br>
<br></p>
<p><input type="radio" name="q2" value="c"> C. It indicates the gene is found in 16 different bacterial species<br>
<br></p>
<p><input type="radio" name="q2" value="d"> D. It estimates species turnover between microbial communities<br>
<br><br></p>
<button type="button" onclick="checkAnswer2()">
<p>Submit</p>
</button>
</form>
<div id="result2" style="margin-top: 1em; font-weight: bold;">

</div>
</div>
<script>
function checkAnswer2() {
  const answer = document.querySelector('#quiz-form-2 input[name="q2"]:checked');
  const result = document.getElementById("result2");

  if (!answer) {
    result.textContent = "⚠️ Please select an answer.";
    result.style.color = "orange";
    return;
  }

  if (answer.value === "b") {
    result.textContent = "✅ Correct! The 'S' in 16S refers to Svedberg units, which describe sedimentation rate during centrifugation.";
    result.style.color = "green";
  } else {
    result.textContent = "❌ Not quite. 16S refers to a sedimentation coefficient, not gene length or diversity metrics.";
    result.style.color = "red";
  }
}
</script>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>